File: 23-06-2024_Longest_Continuous_Subarray_With_Absolute_Diff_Less_Than_or_Equal_to_Limit.py
class Solution:
    def longestSubarray(self, nums: List[int], limit: int) -> int:
        left = 0
        max_len = 0
        increase = deque()
        decrease = deque()
        for right in range(len(nums)):
            while increase and nums[right] < increase[-1]:
                increase.pop()
            increase.append(nums[right])

            while decrease and nums[right] > decrease[-1]:
                decrease.pop()
            decrease.append(nums[right])

            while decrease[0] - increase[0] > limit:
                if nums[left]  == increase[0]:
                    increase.popleft()
                if nums[left] == decrease[0]:
                    decrease.popleft()

                left += 1
            max_len = max(max_len, right - left + 1)
        
        return max_len

File: 23-06-2024_Longest_Continuous_Subarray_With_Absolute_Diff_Less_Than_or_Equal_to_Limit..java
class Solution {
    public int longestSubarray(int[] nums, int limit) {
        LinkedList<Integer> increase = new LinkedList<>();
        LinkedList<Integer> decrease = new LinkedList<>();

        int max = 0,left = 0;
        for(int right = 0; right < nums.length; right++){
            int n = nums[right];

            while(increase.size() > 0 && n < increase.getLast()){
                increase.removeLast();
            }
            increase.add(n);
            while(decrease.size() > 0 && n > decrease.getLast()){
                decrease.removeLast();
            }
            decrease.add(n);

            while(decrease.getFirst() - increase.getFirst() > limit){
                if(nums[left] == decrease.getFirst()){
                    decrease.removeFirst();
                }
                if(nums[left] == increase.getFirst()){
                    increase.removeFirst();
                }
                left++;
            }
            int size = right - left + 1;
            max = Math.max(max, size);
        }
        return  max;
    }
}
